### ISSUE TITLE
Improve configuration struct API

### ISSUE THREAD

**Issue by fho** [rtrlib/rtrlib#4]
- make the defintions of the structs private, this allows to change the internal representation without changing the API
- Copy paramters that are passed to the init(..) functions. The user won't have to care anymore if the passed memory areas are valid while library functions are running
- use errno codes instead of defining our one ones

**Comment by mroethke** [comment]
In response to #86 we discussed required features of a new API for the rtrlib, We came up with following list (without claim to completeness):

- [ ] dynamic group creation and removal
- [ ] dynamic socket creation and removal
- [ ] intervals per socket/cache server
- [ ] memory handling by rtr manager

**Comment by fho** [comment]
Let's resume the discussion about a new API on the base of @nbars suggestion in #86.

- Why do we have rtr_mgr_groups?
They seem to only increase complexity for the user.
Instead of having rtr_mgr_groups, the user could add sockets to the group.
If 2 sockets have the same lowest preference value, the mgr establishes connections to both of them.
If we have distinct cache datasets, we may want to have groups in the internal implementation.
Do Cache server have distinct datasets? E.g. Cache Server A serves all RIPE PFX records, cache server B serves all APNIC PFX records. Or do we have in 99,9% of setups cache-servers that serves all PFX records?

- We should set parameters like  refresh_interval, preference, update_fp in seperate set_ functions for a single parameter. This has the advantages:
  - The parameters can be modified when the manager is running and using the sockets/groups.
  - In most cases the user will be happy with the defaults for those parameters and doesn't use them at all. The function signatures for the _init() functions get shorter and easier to use.

- I like idea from @nbars to provide functions to create/add sockets to a manager from rtr_mgr. I think it is much easier to use than the current approach.
The abstraction for transports can also be removed/reduced and implemented more close in the rtr_socket stuff. It only adds complexity.

- Instead of defining our enum return values, we can use the errno values for errors. They are standardized and in 99% of the cases the available errno values maps fine to the occured problems in rtrlib.

- The packets.c stuff should be reimplemented. It's a totally unreadable mess. :-)

- The abstraction layer for different pfx table implementation is unnecessary. It only makes the code more complex. It's unlikely that we will have in the future a case where we have different pfx table implementation and the user wants to choose between them

- I also added a few comments to https://github.com/rtrlib/rtrlib/pull/86/files

What do you think?

**Comment by fho** [comment]
#38

**Comment by waehlisch** [comment]
@fho thanks for pushing this topic, even though I disagree with some of your remarks ;).

I think this topic should be discussed f2f or in a phone call. Maybe we find time after merging of SPFST/Trie.

**Comment by fho** [comment]
@waehlisch Yes, would be great to discuss the disagreements. I'm not also not sure about some things. :-)
